File: code-road/pickFromHereToMigrate/Arrays-and-Strings/array-and-hashing/Segregate0and1s_solution_1.cpp
#include <iostream>

void oThen1_Solution1(int arr[], int n)
{
  int left = 0, right = n - 1;

  while (left < right)
  {
    while (!(arr[left] & 1) && left < right)
      left++;

    while (arr[right] & 1 && left < right)
      --right;

    if (left < right)
    {
      arr[left] = arr[left] ^ arr[right];
      arr[right] = arr[left] ^ arr[right];
      arr[left] = arr[left] ^ arr[right];

      ++left;
      --right;
    }
  }
}

int main()
{
  // int arr[10] = {0, 1, 0, 1, 0, 1, 0, 1, 0, 1};
  int arr[10] = {1, 1, 1, 1, 1, 1, 1, 1, 1, 0};

  oThen1_Solution1(arr, 10);

  for (int i = 0; i < 10; ++i)
    std::cout << arr[i] << " ";

  std::cout << std::endl;

  return 0;
}

File: rough/pickFromHereToMigrate/Arrays-and-Strings/array-and-hashing/Segregate0and1s_solution_1.js
let arr = [1, 0, 0, 1, 1,];

function segregrate0and1(arr) {
    let left = 0, right = arr.length - 1;

    while (left < right) {
        while (arr[left] == 1 && left < right)
            left++;
        while (arr[right] == 0 && left > right)
            right--;

        if (left < right) {
            arr[left] = 1;
            arr[right] = 0;
            left++;
            right--;
        }
    }

    return arr;
}

console.log('array before', arr);

// since function returned an 'arr', we store it in a newarray 'newarr'
let newarr = segregrate0and1(arr)

console.log('array before', newarr);